### PR TITLE
Fix rllib

### LINKED ISSUE #139: ray[rllib] does not work in godotrl 0.6.0
ray[rllib] 2.2.0 does not currently work in godotrl 0.6.0 as it is still expecting an old gym wrapper but we upgraded our godot env to gymnasium. I tried upgrading to ray[rllib] 2.6.0 (which uses gymnasium instead o gym) but here we have a very minor version missmatch which makes the installation fail. More specfifically `sb3` depends on `gymnasium==0.28.1` and `ray` depends on `gymnasium==0.26.3`.

image

I would love to host ray package on a separate github fork and loosen the gymnasium version to fix this but the ray project seems to be pretty big and a simple pip install from git does not work.

Any ideas how we can tackle this issue? @edbeeching @Ivan-267

### PR DESCRIPTION
**Pull request by edbeeching**
Should resolve #139 

- Updates RLLIB wrappers to support latest version.
- Updates setup to install correct version of gymnasium for rllib
- Make cleanrl work again (use gymnasium instead of gym)
- Make hp tuning work again (use gymnasium instead of gym)

Needs install from new venv with `pip install godot-rl[rllib]`

### CODE REVIEW

**Comment by visuallization** [comment]
@edbeeching haha oh noes! I was working on the same thing as you :D

https://github.com/edbeeching/godot_rl_agents/pull/141

**Comment by edbeeching** [comment]
> @edbeeching haha oh noes! I was working on the same thing as you :D
> 
> #141

Doh!

**Comment by visuallization** [comment]
I think yours is nicer (no check if obs exists and  making infos a list), so we should go with yours. The only thing I noticed is that we apparently have to keep the fixed numpy==1.23.5 version. At least for me it throws an error if we don't keep it.

image

How is it for you? Maybe it is just realted of my godot project which uses an older version of the plugin.

**Comment by visuallization** [comment]
> > @edbeeching haha oh noes! I was working on the same thing as you :D
> > #141
> 
> Doh! Luckily it is not a big change. We should be a bit more organized with assigning things in the future. It seems we made some complementary changes, if you want to pick up the ones from my branch and add them to yours? I will then close mine and we can get yours merged.

Agreed! next time, the one who wants to tackle the problem, should just claim so in the issue.

**Comment by edbeeching** [comment]
> I think yours is nicer (no check if obs exists and making infos a list), so we should go with yours. The only thing I noticed is that we apparently have to keep the fixed numpy==1.23.5 version. At least for me it throws an error if we don't keep it.
> 
> image
> 
> How is it for you? Maybe it is just realted of my godot project which uses an older version of the plugin.

It is strange I don't see to have problem with numpy, but I am running on Linux so perhaps that is it?

I will add your docs changes to my PR.

**Comment by edbeeching** [comment]
@visuallization would you mind installing from scratch with this PR + rllib? I want to be sure that we need to set the numpy version.

**Comment by visuallization** [comment]
> > I think yours is nicer (no check if obs exists and making infos a list), so we should go with yours. The only thing I noticed is that we apparently have to keep the fixed numpy==1.23.5 version. At least for me it throws an error if we don't keep it.
> > image
> > How is it for you? Maybe it is just realted of my godot project which uses an older version of the plugin.
> 
> It is strange I don't see to have problem with numpy, but I am running on Linux so perhaps that is it?
> 
> I will add your docs changes to my PR.

I am also on Linux now - so that's not it. It might be related to godotrl plugin - i will try upgrding the plugin and check again

**Comment by visuallization** [comment]
> @visuallization would you mind installing from scratch with this PR + rllib? I want to be sure that we need to set the numpy version.

I did just that and it is not working for me :(. Not even after using the godot_rl addon from main. which addon version are you using? And what is the numpy version you are using? `pip show numpy`

**Comment by edbeeching** [comment]
> I did just that and it is not working for me :(. Not even after using the godot_rl addon from main. which addon version are you using? And what is the numpy version you are using? `pip show numpy`

```
Name: numpy
Version: 1.25.1
Summary: Fundamental package for array computing in Python
Home-page: https://www.numpy.org
Author: Travis E. Oliphant et al.
Author-email: 
License: BSD-3-Clause
Location: /home/edward/play/godot/godot_rl/godot_rl_agents/venv/lib/python3.10/site-packages
Requires: 
Required-by: contourpy, godot-rl, gym, Gymnasium, huggingface-sb3, imageio, matplotlib, onnx, onnxruntime, pandas, pyarrow, PyWavelets, ray, scikit-image, scipy, stable-baselines3, tensorboard, tensorboardX, tifffile

```

**Comment by visuallization** [comment]
> 

hmm I have python 3.9.16

**Comment by visuallization** [comment]
Hmm maybe it is related on how my actions look.
 2 continuos and 1 discrete. my env is very similar to jumperhard. Does jumperhard work for you with latest numpy version?

**Comment by visuallization** [comment]
@edbeeching Okay so JumperHard does also not working for me with the current numpy version.
Which environment are you using for testing? Is it maybe related to envs which are mixing continuous and discrete spaces?

**Comment by visuallization** [comment]
@edbeeching Okay so yes the actions are exaclty the problem, because the array consists of different types and sizes. see;

image

**Comment by visuallization** [comment]
@edbeeching fixed: https://github.com/edbeeching/godot_rl_agents/pull/140/commits/20c4df7ed73e04eeb04c0e4d16d00091dad7818d

**Comment by Ivan-267** [comment]
> @edbeeching fixed: [20c4df7](https://github.com/edbeeching/godot_rl_agents/commit/20c4df7ed73e04eeb04c0e4d16d00091dad7818d)

Nice catch, this is what I get when trying something similar:
>>> np.array([1,[2,3]])
<stdin>:1: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray with `'1.21.2'` on a local python 3.9 installation. I guess they later changed the warning to an error.

**Comment by visuallization** [comment]
@edbeeching Your implementation does not seem to take num_workers correctly into account. Check this out: Both versions where run with the same config but your implementation took around 20m while mine took around 10m. Not sure what's going on. Rerunning the experiment to make sure I did not mix anything up.

image

**Comment by visuallization** [comment]
> @edbeeching Your implementation does not seem to take num_workers correctly into account. Check this out: Both versions where run with the same config but your implementation took around 20m while mine took around 10m. Not sure what's going on. Rerunning the experiment to make sure I did not mix anything up.
> 
> image

Ah okay it is because with your implementation we have to set --speedup explicitly via terminal

**Comment by visuallization** [comment]
Also fixed some issues with cleanrl and hp tuning

**Comment by visuallization** [comment]
@edbeeching we should probably make a new release for 0.6.1 as it fixes quite a few essential things.

**Comment by edbeeching** [comment]
> @edbeeching we should probably make a new release for 0.6.1 as it fixes quite a few essential things.

Yes I agree, I am just writing a test for rllib training. I will look at #117 then we can look to merge both and release v0.6.1.

**Comment by Ivan-267** [comment]
Would we have to have sb3 as an extra as well so that it doesn't get installed together with rllib?

**Comment by edbeeching** [comment]
@Ivan-267 yes I am trying to avoid that as I would like new users to `pip install godot-rl` and have sb3 installed out of the box. I am looking for a workaround / fix the tests.

**Comment by edbeeching** [comment]
ok while I am glad I made this test. Getting all these dependencies to work is going to be a headache.

**Comment by edbeeching** [comment]
ok a few options:

1. - we remove sb3 as a dependency from the base install and require install with `pip install godot-rl[sb3]`
2. - we wait for https://github.com/ray-project/ray/issues/37393 before adding this fix
3. - we have a more convoluted install process `pip install godot-rl`, `pip uninstall sb3`, `pip install ray[rllib]`

I actually prefer the third option for now, as it will mean we can test correctly and new users have sb3 which is the best entrypoint to the lib. Advanced users can follow the steps above for rllib support. What do you think @visuallization ?

File: .github/workflows/test-ci.yml [hunk 1]
```diff
@@ -55,3 +55,57 @@ jobs:
...
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        # cpu version of pytorch
+        pip install .[test]
+        pip uninstall -y stable-baselines3 gymnasium
+        pip install -y ray[rllib]
```
Comment (Ivan-267) [inline .github/workflows/test-ci.yml:78]:
```suggestion
        pip install ray[rllib]
```

File: .github/workflows/test-ci.yml [hunk 2]
```diff
@@ -55,3 +55,57 @@ jobs:
...
```
Comment (Ivan-267) [inline .github/workflows/test-ci.yml:104]:
```suggestion
        pip install ray[rllib]
```

**Comment by Ivan-267** [comment]
Actually even with those changes I got:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
godot-rl 0.6.1 requires stable-baselines3, which is not installed.
```

**Comment by visuallization** [comment]
> ok a few options:
> 
>     1. * we remove sb3 as a dependency from the base install and require install with `pip install godot-rl[sb3]`
> 
>     2. * we wait for [[RLlib] Upgrade the gymnasium dependency to version the latest version (gymnasium==0.28.1) ray-project/ray#37393](https://github.com/ray-project/ray/issues/37393) before adding this fix
> 
>     3. * we have a more convoluted install process `pip install godot-rl`, `pip uninstall sb3`, `pip install ray[rllib]`
> 
> 
> I actually prefer the third option for now, as it will mean we can test correctly and new users have sb3 which is the best entrypoint to the lib. Advanced users can follow the steps above for rllib support. What do you think @visuallization ?

You mean for testing? I think the 3rd option is fine. I would love option number 2 but it might still take some time until it gets released (if it ever does) - so let's go for now with option 3, as I think it it is important to fix the current issues and we can still  go with option 2, once ray releases the proposed upgrade. I also don't like option 1 too much because of the points you mentioned - it should be a easy as possible to get started with godotrl-agents.

On another note: I was wondering if we should fix our dependency versions? With that we can make sure that things don't randomly break if somebody installs the lib at a later point. We would then need to have a dedicated updated process, where we go through the dependencies every now and then, test things, and fix a new version. I think it might be actually worth it in terms of stabiility and future proof. @edbeeching @Ivan-267 what do you guys think?

**Comment by visuallization** [comment]
@edbeeching I fixed the install step in the tests but now the actual test seems to timeout for rllib.

Check this regarding fixing wheel version :shrug: https://github.com/freqtrade/freqtrade/issues/8376

**Comment by Ivan-267** [comment]
About fixing versions, I also think that it would help with stability, but then we may have to make sure that the version (or version ranges) specified are compatible for most operating systems / Python versions that we want to support. Maybe that's the case for all or most packages, just something to consider.

The last test yesterday also did start successfully but was taking too long.

**Comment by Ivan-267** [comment]
> > The last test yesterday also did start successfully but was taking too long
> 
> @Ivan-267 Oh really? Maybe I misslooked but it seemed like it would still throw the same pip install error Ed reported. Anyways if your version works, please commit. :)

I accidentally made it as two commits (from single comment two suggestions), but the second one did pass:

It just shows the same error message as I got locally (but it doesn't halt the test):
https://github.com/edbeeching/godot_rl_agents/actions/runs/5685626539/job/15422792046
```
godot-rl 0.6.1 requires stable-baselines3, which is not installed.
Successfully installed PyWavelets-1.4.1 aiosignal-1.3.1 attrs-23.1.0 click-8.1.6 dm-tree-0.1.8 frozenlist-1.4.0 gymnasium-0.26.3 gymnasium-notices-0.0.1 imageio-2.31.1 jsonschema-4.18.4 jsonschema-specifications-2023.7.1 lazy_loader-0.3 lz4-4.3.2 markdown-it-py-3.0.0 mdurl-0.1.2 msgpack-1.0.5 pyarrow-12.0.1 pygments-2.15.1 ray-2.6.1 referencing-0.30.0 rich-13.4.2 rpds-py-0.9.2 scikit-image-0.21.0 scipy-1.11.1 tensorboardX-2.6.1 tifffile-2023.7.18 typer-0.9.0
```

However I wouldn't say it works because the test is still taking too long as with your case. I'm not sure why, we could try changing some config and see if it makes a difference.

I was trying the test itself locally with pytest, but I found an issue with the example download step (for some reason it's not cloning the large files, e.g. the exe for me). However my local Windows environment may not be exactly the same as the test environment (or maybe I missed some step), as I see the test env downloads the examples just fine.

The passed test with warnings above was from `pytest .\tests\test_rllib.py` on my local system (after rebuilding the exe).

**Comment by Ivan-267** [comment]
It looks like it is passing now (similarly to on my PC, with warnings), well at least on the two platforms that completed the test so far. 

I'm not sure if it was the timesteps, changing num workers to 1 (maybe the multiple workers cause some issue with test env? it works OK on my local setup), or some of the other changes that fixed it.

**Comment by edbeeching** [comment]
Hey thanks to you both for continuing to work on this. I am travelling this weekend and will pick up on Monday.

**Comment by visuallization** [comment]
> It looks like it is passing the rllib test now (similarly to as on my PC, with warnings).
> 
> I'm not sure if it was the timesteps, changing num workers to 1 (maybe the multiple workers cause some issue with test env? it works OK on my local setup), or some of the other changes that fixed it.
> 
> You could check the remaining error:
> 
> ```
> godot_rl/wrappers/sample_factory_wrapper.py:100: AttributeError
> =========================== short test summary info ============================
> FAILED tests/test_sample_factory.py::test_sample_factory_training - AttributeError: 'Namespace' object has no attribute 'seed'
> ```

@Ivan-267 that's awesome! Thanks for fixing the tests including the seeding issue - I forgot about the sample factory example env!

**Comment by edbeeching** [comment]
@visuallization or @Ivan-267 thanks for fixing the tests. I just pushed an update to the rllib installation instructions. If you can approve the PR then I will merge.

**Comment by visuallization** [comment]
@edbeeching @Ivan-267 Ray is updating gymnasium: https://github.com/ray-project/ray/issues/37393#issuecomment-1660369444. :tada: Will be release in around 3 weeks. So  we could keep an `all` option and fix the version for it, so it stays compatible in the future.

**Comment by edbeeching** [comment]
@visuallization , this is great. I will merge as is for now as there are a number of other fixes in this PR.

**Pull request merged**